src/include/replication/replication_manager.h

```diff
@@ -0,0 +1,204 @@
+
+  /** @return   True if this is the primary node and false if this is a replica. */
+  bool IsPrimary() const { return identity_ == "primary"; }
+
+ private:
+  /** The main event loop that all nodes run. This handles receiving messages. */
+  void EventLoop(common::ManagedPointer<messenger::Messenger> messenger, const messenger::ZmqMessage &msg);
+
+  /**
+   * Request a heartbeat from the specified replica.
+   * Note that this function does not wait for a response and will update the replica's last heartbeat time in the
+   * background. This means that ReplicaHeartbeat may need to be invoked again to realize that the replica is dead.
+   *
+   * If REPLICATION_CARDIAC_ARREST_MS seconds have passed since the last acknowledgement, the replica is marked dead.
+   *
+   * @param replica_name                The replica to request a heartbeat from.
+   */
+  void ReplicaHeartbeat(const std::string &replica_name);
+
+  /**
+   * Build the list of replicas as specified in replication.config.
+   *
+   * @param replication_hosts_path      The path to the replication.config file.
+   */
+  void BuildReplicaList(const std::string &replication_hosts_path);
+
+  /**
+   * Connect to the specified replica.
+   *
+   * @param replica_name                The name to assign the replica.
+   * @param hostname                    The hostname where the replica is available.
+   * @param port                        The port where the replica is available.
+   */
+  void ReplicaConnect(const std::string &replica_name, const std::string &hostname, uint16_t port);
+
+  /** @return The connection ID associated with a particular replica. */
+  common::ManagedPointer<messenger::ConnectionId> GetReplicaConnection(const std::string &replica_name);
+
+  common::ManagedPointer<messenger::Messenger> messenger_;  ///< The messenger used for all send/receive operations.
+  std::string identity_;                                    ///< The identity of this replica.
+  uint16_t port_;                                           ///< The port that replication runs on.
+  std::unordered_map<std::string, Replica> replicas_;       ///< Replica Name -> Connection ID.
+
+  std::unique_ptr<storage::ReplicationLogProvider> provider_;  ///< The replicated buffers provided by the primary.
+  std::mutex blocking_send_mutex_;                             ///< Mutex used for blocking sends.
+  std::condition_variable blocking_send_cvar_;                 ///< Cvar used for blocking sends.
+
+  // TODO(WAN): I think it is better to use retention policies instead of enabling/disabling replication.
+  bool replication_enabled_ = false;  ///< True if replication is currently enabled and false otherwise.
+
+  uint64_t next_buffer_sent_id_ = 1;      ///< The ID of the next buffer to sent.
+  uint64_t last_record_received_id_ = 0;  ///< The ID of the last record to be received.
+
+  /** The received messages are queued up until the "next" received message is the right one. */
```

Comment:
May want to include what "right one" means

Reply:
Added (I hope).